supervisor/options.py

```diff
@@ -1939,6 +1952,57 @@ def make_dispatchers(self, proc):
             dispatchers[stdin_fd] = PInputDispatcher(proc, 'stdin', stdin_fd)
         return dispatchers, p
 
+    def load_external_environment_definition(self):
+        return self.load_external_environment_definition_for_config(self)
+
+    # this is separated out in order to make it easier to test
+    @classmethod
+    def load_external_environment_definition_for_config(cls, config):
+        # lazily load extra env vars before we drop privileges so that this can be used to load a secrets file
+        # or execute a program to get more env configuration. It doesn't have to be secrets, just config that
+        # needs to be separate from the supervisor config for whatever reason. The supervisor config interpolation
+        # is not supported here. The data format is just plain text, with one k=v value per line. Lines starting
+        # with '#' are ignored.
+        env = {}
+        envdata = None
+        if config.environment_file:
+            if os.path.exists(config.environment_file):
+                try:
+                    with open(config.environment_file, 'r') as f:
+                        envdata = f.read()
+
+                except Exception as e:
+                    raise ProcessException("environment_file read failure on %s: %s" % (config.environment_file, e))
+
+        elif config.environment_loader:
+            try:
+                from subprocess import check_output, CalledProcessError
+                kwargs = dict(shell=True)
+                if not PY2:
+                    kwargs['text'] = True
+
+                envdata = check_output(config.environment_loader, **kwargs)
```

Comment:
Good point - I did put that in the wrong spot. I had originally put that into the _spawn_as_child method, but then got lost in the weeds trying to figure out the tests. I've updated this with a change to make this call in the child process in _spawn_as_child, after the fork but before any process modifications. This will ensure the main process is not blocked, and the optional envirnoment configuration only affects the startup time of the configured processes.

Seems better to me.

Reply:
Good point. I put that call in the wrong place and have moved it into the _spawn_as_child method and added a scary comment on the implementation. I had initially done it that way but got lost in the weeks trying to figure out the tests. This is much better and doesn't affect the main process at all.

Reply:
> I've updated this with a change to make this call in the child process in _spawn_as_child, after the fork

Putting it after the fork addresses my concern above.  So far I've only done a quick scan of the diff, but that line stood out to me as a potential issue.

Reply:
Thanks for taking the time to review and consider this, appreciate your feedback.